Improve Hubble memory usage and performance on decoding events

tklauser opened the pull request:
The first 7 commits are preparatory cleanup and addition/extension of tests to avoid regressions introduced by the successive two commits. These reduce Hubble's memory usage on L3/4 and L7 event decoding as follows:

    name         old time/op    new time/op    delta
    L34Decode-8    4.10µs ± 5%    3.63µs ± 8%  -11.51%  (p=0.000 n=10+10)
    
    name         old alloc/op   new alloc/op   delta
    L34Decode-8    1.24kB ± 0%    1.03kB ± 0%  -16.77%  (p=0.000 n=10+10)
    
    name         old allocs/op  new allocs/op  delta
    L34Decode-8      26.0 ± 0%      21.0 ± 0%  -19.23%  (p=0.000 n=10+10)
    
    name        old time/op    new time/op    delta
    L7Decode-8    6.37µs ±11%    5.79µs ± 4%   -9.02%  (p=0.000 n=10+10)
    
    name        old alloc/op   new alloc/op   delta
    L7Decode-8    1.53kB ± 0%    1.37kB ± 0%  -10.47%  (p=0.000 n=10+10)
    
    name        old allocs/op  new allocs/op  delta
    L7Decode-8      36.0 ± 0%      34.0 ± 0%   -5.56%  (p=0.000 n=10+10)

See the individual commit messages for details.

tklauser committed:
hubble: remove duplicate import in agent event parser tests

Found using staticcheck.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
hubble/parser/threefour: test Decode on empty flow

To avoid regressions in successive changes.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
hubble: omit redundant nil check on slices

len is defined for all slices, even nil slices. These have a length of
0, so these checks are not necessary.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
labels: preallocate maps and avoid using json.Decoder

Preallocate maps with known size and call json.Unmarshal directly, no
need to allocate a json.Decoder.

Suggested-by: Robin Hahling <robin.hahling@gw-computing.net>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

rolinh commented on pkg/labels/labels.go:
```diff
@@ -269,6 +267,7 @@ func (l *Label) UnmarshalJSON(data []byte) error {
 	}
 
+	decoder := json.NewDecoder(bytes.NewReader(data))
 	err := decoder.Decode(&aux)
```
I think we can avoid using a decoder and directly attempt to unmarshal, no?
```suggestion
	err := json.Unmarshal(data, &aux)
```
Same comment applies for the other instance a few lines below.

tklauser replied:
Good point, I'll fold that change into the commit touching `pkg/labels`.

rolinh reviewed (commented):
LGTM overall although I have one little suggestion, see below.

tklauser committed:
identity/model: preallocate slice and map with known size

Found using prealloc.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
monitor: remove unused functions

These are unused since commit b88afa221afc ("monitor: new listener v1.2
with reusing en/decoder").

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
monitor: add test for DecodeTraceNotify

To avoid regressions in successive changes to DecodeTraceNotify.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
monitor: don't allocate on reading TraceNotify version

Use byteorder.Native.Uint16 to read the version uint16 directly instead
of allocating a bytes.Reader. For L34Decode in hubble this improves
performance and memory for each decoded TraceNotify event as follows:

name         old time/op    new time/op    delta
L34Decode-8    4.30µs ± 7%    4.00µs ± 5%   -6.96%  (p=0.000 n=10+10)

name         old alloc/op   new alloc/op   delta
L34Decode-8    1.24kB ± 0%    1.19kB ± 0%   -3.87%  (p=0.000 n=10+10)

name         old allocs/op  new allocs/op  delta
L34Decode-8      26.0 ± 0%      23.0 ± 0%  -11.54%  (p=0.000 n=10+10)

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
hubble: let ServiceGetter.GetServiceByAddr return service by pointer

This avoids copying the Service protobuf object and improves performance
and memory usage:

name         old time/op    new time/op    delta
L34Decode-8    4.10µs ± 5%    3.63µs ± 8%  -11.51%  (p=0.000 n=10+10)

name         old alloc/op   new alloc/op   delta
L34Decode-8    1.24kB ± 0%    1.03kB ± 0%  -16.77%  (p=0.000 n=10+10)

name         old allocs/op  new allocs/op  delta
L34Decode-8      26.0 ± 0%      21.0 ± 0%  -19.23%  (p=0.000 n=10+10)

name        old time/op    new time/op    delta
L7Decode-8    6.37µs ±11%    5.79µs ± 4%   -9.02%  (p=0.000 n=10+10)

name        old alloc/op   new alloc/op   delta
L7Decode-8    1.53kB ± 0%    1.37kB ± 0%  -10.47%  (p=0.000 n=10+10)

name        old allocs/op  new allocs/op  delta
L7Decode-8      36.0 ± 0%      34.0 ± 0%   -5.56%  (p=0.000 n=10+10)

Also, the ok bool return is no longer necessary as callers can just
check for a nil return value. This simplifies the logic in
(*Parser).Decode a bit and avoids allocating an empty object for cases
when no service is found.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

kkourt commented on pkg/identity/model/identity.go:
```diff
@@ -16,7 +16,7 @@ func NewIdentityFromModel(base *models.Identity) *identity.Identity {
 	id := &identity.Identity{
 		ID:     identity.NumericIdentity(base.ID),
-		Labels: make(labels.Labels),
+		Labels: make(labels.Labels, len(base.Labels)),
```
Any chance this needs to be `make(labels.Labels, 0, len(base.Labels))` instead?

tklauser replied:
That field has type `labels.Labels` which is defined as a map type:

https://github.com/cilium/cilium/blob/0666f5355de975733e8d1b094aade7fd3a6b8786/pkg/labels/labels.go#L113

`make` for maps only takes an optional size argument, but no capacity according to https://pkg.go.dev/builtin#make

kkourt replied:
Ah makes sense! Thanks!

This was on the `preallocate slices` commit rather than the `preallocate maps` one which is why I wasn't sure about above.

tklauser replied:
Ah, I see. That commit message is misleading. It's preallocating a slice and a map. Thanks for pointing that out! I'll change it accordingly.

kkourt reviewed (approved):
🚀 

Just a minor comment to a change that seemed a bit odd to me :)
